pkg/op/device.go

```diff
@@ -0,0 +1,291 @@
+				WithDescription(fmt.Sprintf("required client auth method: %s", m))
+		}
+	}
+
+	resp, err := CreateTokenResponse(ctx, authReq, client, exchanger, true, state.AuthCode, "")
+	if err != nil {
+		return err
+	}
+
+	httphelper.MarshalJSON(w, resp)
+	return nil
+}
+
+func ParseDeviceAccessTokenRequest(r *http.Request, exchanger Exchanger) (*oidc.DeviceAccessTokenRequest, error) {
+	req := new(oidc.DeviceAccessTokenRequest)
+	if err := exchanger.Decoder().Decode(req, r.PostForm); err != nil {
+		return nil, err
+	}
+	return req, nil
+}
+
+func CheckDeviceAuthorizationState(ctx context.Context, clientID, deviceCode string, exchanger Exchanger) (*DeviceAuthorizationState, AuthRequest, error) {
+	storage, err := assertDeviceStorage(exchanger.Storage())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	state, err := storage.GetDeviceAuthorizatonState(ctx, clientID, deviceCode)
+	if errors.Is(err, context.DeadlineExceeded) {
+		return nil, nil, oidc.ErrSlowDown().WithParent(err)
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+	if state.Denied {
+		return state, nil, oidc.ErrAccessDenied()
+	}
+	if state.AuthCode != "" {
+		return state, nil, nil
+	}
+	if time.Now().After(state.Expires) {
+		return state, nil, oidc.ErrExpiredDeviceCode()
+	}
+	authReq, err := AuthRequestByCode(ctx, exchanger.Storage(), state.AuthCode)
+	return state, authReq, err
+}
+
+func userCodeVerificationHandler(o OpenIDProvider) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		UserCodeVerification(w, r, o)
+	}
+}
+
+type UserCodeVerificationRequest struct {
+	Code        string `schema:"code"`
+	UserCode    string `schema:"user_code"`
+	RedirectURL string `schema:"redirect_url"`
+}
+
+func UserCodeVerification(w http.ResponseWriter, r *http.Request, o OpenIDProvider) {
+	if err := userCodeVerification(w, r, o); err != nil {
+		RequestError(w, r, err)
+	}
+}
+
+func userCodeVerification(w http.ResponseWriter, r *http.Request, o OpenIDProvider) (err error) {
+	req, err := ParseUserCodeVerificationRequest(r, o.Decoder())
+	if err != nil {
+		return err
+	}
+
+	storage, err := assertDeviceStorage(o.Storage())
+	if err != nil {
+		return err
+	}
+
+	ctx := r.Context()
+	if err := storage.CompleteDeviceAuthorization(ctx, req.Code, req.UserCode); err != nil {
+		return err
+	}
+
+	if req.RedirectURL != "" {
+		http.Redirect(w, r, req.RedirectURL, http.StatusSeeOther)
+	}
+
+	fmt.Fprintln(w, "Authorization successfull, please return to your device")
+	return nil
+}
+
+func ParseUserCodeVerificationRequest(r *http.Request, decoder httphelper.Decoder) (*UserCodeVerificationRequest, error) {
+	if err := r.ParseForm(); err != nil {
+		return nil, oidc.ErrInvalidRequest().WithDescription("cannot parse form").WithParent(err)
+	}
+
+	req := new(UserCodeVerificationRequest)
+	if err := decoder.Decode(req, r.Form); err != nil {
+		return nil, oidc.ErrInvalidRequest().WithDescription("cannot parse user code form").WithParent(err)
+	}
+	if req.Code == "" {
```

Comment:
What information has to be in "Code" as in the RFC are only user_code and the verification_uri mentioned? If it's not necessary or a secondary attribute from user_code then the request will fail

Reply:
as discussed, handler removed